platform/jewel/markdown/core/src/main/kotlin/org/jetbrains/jewel/markdown/rendering/ImageSourceResolver.kt

```diff
@@ -27,35 +29,121 @@ public interface ImageSourceResolver {
+         * - plain URIs, e.g., `https://example.com/image.png`
+         * - relative paths in the current classloader's resources, e.g., `/images/my-image.png`
+         * - relative paths relative to a given root directory [rootDir], e.g., `../images/my-image.png`
+         *
+         * If [logResolveFailure] is true, logs any failures to resolve image sources.
+         */
+        public fun create(rootDir: Path, logResolveFailure: Boolean): ImageSourceResolver =
+            create(
+                buildList {
+                    addAll(defaultCapabilities)
+                    add(ResolveCapability.RelativePath(rootDir))
+                },
+                logResolveFailure,
+            )
+
+        /**
+         * Creates [ImageSourceResolver] that can resolve image links in Markdown files according to provided
+         * [resolveCapabilities].
+         *
+         * If [logResolveFailure] is true, logs any failures to resolve image sources.
+         */
+        public fun create(
+            resolveCapabilities: List<ResolveCapability> = defaultCapabilities,
+            logResolveFailure: Boolean = true,
+        ): ImageSourceResolver = DefaultImageSourceResolver(resolveCapabilities)
+    }
+
+    /** Provides a list of capabilities that the default [ImageSourceResolver] implementation supports. */
+    @ApiStatus.Experimental
+    @ExperimentalJewelApi
+    public sealed interface ResolveCapability {
+        public val debugName: String
+
+        /** Resolves a raw image destination string from a Markdown file into a fully-qualified, loadable path. */
+        public fun resolve(rawDestination: String): String?
+
+        /** Represents the ability to resolve plain URIs as-is. */
+        @ApiStatus.Experimental
+        @ExperimentalJewelApi
+        public object PlainUri : ResolveCapability {
+            override val debugName: String
+                get() = "PlainUri"
+
+            override fun resolve(rawDestination: String): String? {
+                val uri = runCatching { URI.create(rawDestination) }.getOrNull() ?: return null
+                return if (uri.isAbsolute) rawDestination else null
+            }
+        }
+
+        /** Represents the ability to resolve relative paths in the current classloader's resources. */
+        @ApiStatus.Experimental
+        @ExperimentalJewelApi
+        public object RelativePathInResources : ResolveCapability {
+            override val debugName: String
+                get() = "RelativePathInResources"
+
+            override fun resolve(rawDestination: String): String? =
+                javaClass.classLoader.getResource(rawDestination.removePrefix("/"))?.toExternalForm()
+        }
+
+        /** Represents the ability to resolve relative paths relative to a given root directory [rootDir]. */
+        @ApiStatus.Experimental
+        @ExperimentalJewelApi
+        public class RelativePath(private val rootDir: Path) : ResolveCapability {
```

Comment:
Would this work with an absolute file path, too? E.g., `/Users/seb/Downloads/image.png` or `C:\temp\picture.jpg`

If not, is it something we can support?

Reply:
Added support for absolute paths